modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -113,8 +147,174 @@ public SnapshotReader getReader() {
+                    MSG_FACTORY.snapshotMvDataRequest().id(snapshotUri.snapshotId).build(),
+                    NETWORK_TIMEOUT
+            ).whenComplete((networkMessage, throwable) -> {
+                if (throwable != null) {
+                    future.completeExceptionally(copierException(throwable, RaftError.EIO, "Failed to load mv partition data"));
+                } else {
+                    SnapshotMvDataResponse snapshotMvDataResponse = ((SnapshotMvDataResponse) networkMessage);
+
+                    handleVersionChains(snapshotMvDataResponse.rows());
+
+                    if (!snapshotMvDataResponse.finish()) {
+                        requestSnapshotMvDataAsync(snapshotSenderNode);
+                    } else {
+                        completeCopySnapshot();
+                    }
+                }
+            });
+        } catch (Throwable t) {
+            future.completeExceptionally(t);
+        } finally {
+            cancelBusyLock.leaveBusy();
+        }
+    }
+
+    private void completeCopySnapshot() {
+        if (!cancelBusyLock.enterBusy()) {
+            return;
+        }
+
+        try {
+            SnapshotMeta snapshotMeta0 = snapshotMeta;
+
+            assert snapshotMeta0 != null;
+
+            partitionSnapshotStorage.partition().lastAppliedIndex(snapshotMeta0.lastIncludedIndex());
+
+            future.complete(null);
+        } catch (StorageException e) {
+            future.completeExceptionally(copierException(e, RaftError.EIO, "Error writing last applied index"));
+        } catch (Throwable t) {
+            future.completeExceptionally(t);
+        } finally {
+            cancelBusyLock.leaveBusy();
+        }
+    }
+
+    private void handleVersionChains(List<ResponseEntry> responseEntries) {
+        for (ResponseEntry entry : responseEntries) {
+            if (!cancelBusyLock.enterBusy()) {
+                return;
+            }
+
+            try {
+                partitionSnapshotStorage.partition().writeVersionChain(
+                        entry.rowId(),
+                        entry.rowVersions(),
+                        entry.timestamps(),
+                        entry.txId(),
+                        entry.commitTableId(),
+                        entry.commitPartitionId()
+                );
+            } catch (StorageException e) {
+                future.completeExceptionally(copierException(e, RaftError.EIO, "Error writing version chain"));
+            } catch (Throwable t) {
+                future.completeExceptionally(t);
+            } finally {
+                cancelBusyLock.leaveBusy();
+            }
+        }
+    }
+
+    /**
+     * Exception along with its status for snapshot copier.
+     *
+     * <p>The class is needed because the copier status is not thread-safe.
+     *
+     * @see #isOk()
+     * @see #setError(RaftError, String, Object...)
+     */
+    static class SnapshotCopierException extends RuntimeException {
```

Comment:
This is weird. Not even a StorageException or anything. We should start thinking in terms of error groups and such